File: NeuronEngine OLD CS/NeuronBase.cs
using System;
using System.Collections.Generic;
using System.Threading;
namespace BrainSimulator
{
    public class NeuronBase
    {
        public enum baseModelType { Std, Color, FloatValue, LIF, Random };
        public const int threshold = 10000;
        public const float scaleFactor = (float)threshold;
        public long lastFired = 0;
        int id = -1;
        public int Id { get => id; set => id = value; }

        //the accumulating value of a neuron
        public int currentCharge = 0;

        public baseModelType model = baseModelType.Std;
        public baseModelType Model { get => model; set => model = value; }
        public string label = ""; //for convenience only...not used in computation
        public string Label { get { return label; } set { label = value; } }
        public bool keepHistory = false;
        //This is the way to set up a list so it saves and loads properly from an XML file
        public List<SynapseBase> synapses;// = new List<Synapse>();
        public List<SynapseBase> synapsesFrom;// = new List<Synapse>();
        public int lastCharge = 0;  //the ending value of a neuron (possibly after reset)
        public float LastCharge { get { return (float)lastCharge / NeuronBase.scaleFactor; } set { lastCharge = (int)(value * NeuronBase.scaleFactor); } }
        public int LastChargeInt { get { return lastCharge; } set { lastCharge = value; } }
        protected bool alreadyInQueue = false; //used by the queue-based engine 
        protected int alreadyProcessedSynapses = 0;
        private long nextFiring = 0;
        public float LeakRate = 0.1f; //used only by LIF model
        public bool InUse()
        {
            return ((synapses != null && synapses.Count != 0) || (synapsesFrom != null && synapsesFrom.Count != 0) || Label != "");
        }
        public NeuronBase()
        {
            synapses = new List<SynapseBase>();
            synapsesFrom = new List<SynapseBase>();
            Model = baseModelType.Std;
        }
        public NeuronBase(bool allocateSynapses = false)
        {
            if (allocateSynapses)
            {
                synapses = new List<SynapseBase>();
                synapsesFrom = new List<SynapseBase>();
            }
            Model = baseModelType.Std;
        }
        public NeuronBase(int id1, baseModelType t = baseModelType.Std)
        {
            Id = id1;
            Model = t;
        }
        public void Fire1(NeuronArrayBase theNeuronArray, List<int> nextQueue, List<int> zeroQueue)
        {
            NeuronBase[] neuronArray = theNeuronArray.baseNeuronArray;
            if (lastCharge < threshold) return;
            if (Interlocked.CompareExchange(ref alreadyProcessedSynapses, 1, 0) == 1) return;    //this prevents duplicates in the firing queue from being processed twice
            if (synapses != null)
                for (int i = 0; i < synapses.Count; i++) //process all the synapses sourced by this neuron
                {
                    SynapseBase s = synapses[i];
                    NeuronBase n = s.N;
                    //Interlocked.Add(ref n.currentCharge, s.IWeight);
                    n.currentCharge += s.IWeight;
                    //if the target neuron needs processing, add it to the firing queue
                    if (!n.alreadyInQueue)
                    {
                        if (n.currentCharge >= threshold)// || n.currentCharge < 0))
                        {
                            n.alreadyInQueue = true; //a few duplicates are no problem because of the alreadyprocessed flag
                            nextQueue.Add(s.N.Id);  //we don't have to lock because this tread owns the queue
                        }
                    }
                    if (n.currentCharge < 0)
                    {
                        zeroQueue.Add(s.N.Id);
                    }
                }
        }
        public void Fire2(NeuronArrayBase theNeuronArray, List<int> firedQueue)
        {//check for firing
            alreadyInQueue = false;
            alreadyProcessedSynapses = 0;
            if (currentCharge < 0) currentCharge = 0;
            lastCharge = currentCharge;
            if (lastCharge >= threshold)
            {
                currentCharge = 0;
                lastFired = theNeuronArray.Generation;
                //add yourself to the firing queue for next time to reset lastCharge on the next generation
                firedQueue.Add(Id);
                alreadyInQueue = true;
            }
        }
    }
}

File: CppSpeedTest - Copy/NeuronBase.cpp
#include "NeuronBase.h"
#include "SynapseBase.h"
#include "NeuronArrayBase.h"

namespace NeuronEngine
{
	NeuronBase::NeuronBase(int ID)
	{
		leakRate = 0;
		nextFiring = 0;
		id = ID;
	}
	
	NeuronBase::~NeuronBase()
	{
		if (synapses != NULL)
		{
			delete synapses;
		}
	}

	int NeuronBase::GetId() 
	{
		return id;
	}


	NeuronBase::modelType NeuronBase::GetModel() 
	{
		return model;
	}

	void NeuronBase::SetModel(modelType value)
	{
		model = value;
	}

	float NeuronBase::GetLastCharge()
	{
		return lastCharge;
	}
	void NeuronBase::SetLastCharge(float value)
	{
		lastCharge = value;
		currentCharge = value;
	}
	float NeuronBase::GetCurrentCharge()
	{
		return currentCharge;
	}
	float NeuronBase::GetLeakRate()
	{
		return leakRate;
	}
	void NeuronBase::SetLeakRate(float value)
	{
		leakRate = value;
	}
	long NeuronBase::GetLastFired()
	{
		return lastFired;
	}

	void NeuronBase::AddSynapse(NeuronBase* n, float weight, bool isHebbian)
	{
		SynapseBase s;
		s.SetWeight(weight);
		s.SetTarget(n);
		s.SetIsHebbian(isHebbian);
		if (synapses == NULL)
		{
			synapses = new std::vector<SynapseBase>();
		}

		for (int i = 0; i < synapses->size(); i++)
		{
			if (synapses->at(i).GetTarget() == n)
			{
				synapses->at(i).SetWeight(synapses->at(i).GetWeight() + weight);
				goto alreadyInList;
			}
		}
		synapses->push_back(s);
	alreadyInList:
		return;
	}

	void NeuronBase::DeleteSynapse(NeuronBase* n, float weight)
	{
		for (int i = 0; i < synapses->size(); i++)
		{
			if (synapses->at(i).GetTarget() == n)
			{
				if (weight == 0)
					synapses->erase(synapses->begin() + i);
				else
				{
					synapses->at(i).AddToWeight(-weight);
					if (synapses->at(i).GetWeight() == 0)
						synapses->erase(synapses->begin() + i);
				}
				break;
			}
		}
		if (synapses->size() == 0)
		{
			delete synapses;
			synapses = NULL;
		}
	}
	std::vector<SynapseBase>* NeuronBase::GetSynapses()
	{
		return synapses;
	}


	void NeuronBase::Fire1NoQ()
	{
		if (lastCharge < threshold)return;
		if (synapses != NULL)
		{
			for (long i = 0; i < synapses->size(); i++) //process all the synapses sourced by this neuron
			{
				SynapseBase s = synapses->at(i);
				NeuronBase* n = s.GetTarget();
				n->currentCharge = n->currentCharge + s.GetWeight();
				if (s.IsHebbian())
				{
					if (n->currentCharge > threshold)
						s.AddToWeight(.001);
					else
						s.AddToWeight(-.0001);
				}
			}
		}
	}

	bool NeuronBase::Fire2NoQ(long generation)
	{ //check for firing
		if (currentCharge < 0)currentCharge = 0;
		lastCharge = currentCharge;
		if (lastCharge >= threshold) {
			lastFired = generation;
			currentCharge = 0;
			return true;
		}
		return false;
	}



	//void NeuronBase::Fire1(std::vector<NeuronBase*> &neuronArray, std::vector<int>& nextQueue, std::vector<int>& zeroQueue)
	//{
	//	if (lastCharge < threshold)return;
	//	if (alreadyProcessedSynapses) return; //this prevents duplicates in the firing queue from being processed twice
	//	alreadyProcessedSynapses = true;
	//	for (int i = 0; i < synapses.size(); i++) //process all the synapses sourced by this neuron
	//	{
	//		SynapseBase* s = synapses[i];
	//		NeuronBase* n = s->N;
	//		//Interlocked.Add(ref n.currentCharge, s.IWeight);
	//		n->currentCharge += s->IWeight;
	//		//if the target neuron needs processing, add it to the firing queue
	//		if (!n->alreadyInQueue)
	//		{
	//			if (n->currentCharge >= threshold) // || n.currentCharge < 0))
	//			{
	//				n->alreadyInQueue = true; //a few duplicates are no problem because of the alreadyprocessed flag
	//				nextQueue.push_back(n->getId()); //we don't have to lock because this tread owns the queue
	//			}
	//		}
	//		if (n->currentCharge < 0)
	//		{
	//			zeroQueue.push_back(s->N->getId());
	//		}
	//	}
	//}

	//void NeuronBase::Fire2(std::vector<int>& firedQueue)
	//{ //check for firing
	//	alreadyInQueue = false;
	//	alreadyProcessedSynapses = 0;
	//	if (currentCharge < 0)
	//	{
	//		currentCharge = 0;
	//	}
	//	lastCharge = currentCharge;
	//	if (lastCharge >= threshold)
	//	{
	//		currentCharge = 0;
	//		//add yourself to the firing queue for next time to reset lastCharge on the next generation
	//		firedQueue.push_back(getId());
	//		alreadyInQueue = true;
	//	}
	//}


}
